internal/controller/pipelinerun_controller_test.go

```diff
@@ -14,19 +14,119 @@ See the License for the specific language governing permissions and
+			pipelineRuns := listPipelineRuns(MintMakerNamespaceName)
+			for _, pipelinerun := range pipelineRuns {
+				Expect(k8sClient.Delete(ctx, &pipelinerun)).Should(Succeed())
+			}
+		})
+
+		It("should successfully launch new pipelineruns", func() {
+
+			existingPipelineRuns := tektonv1.PipelineRunList{
+				Items: listPipelineRuns(MintMakerNamespaceName),
+			}
+			// To count running pipelineruns, we have to check for empty 'status' fields;
+			// During the tests, these pipelineruns will not be updated to be actually 'running'
+			count := 0
+			for _, pipelineRun := range existingPipelineRuns.Items {
+				if pipelineRun.Spec.Status == "" {
+					count += 1
+				}
+			}
+			Expect(count).Should(Equal(MaxSimultaneousPipelineRuns))
+		})
+	})
+
+	Context("When launching new pipelineruns", func() {
+
+		originalMaxSimultaneousPipelineRuns := MaxSimultaneousPipelineRuns
+
+		_ = BeforeEach(func() {
+			createNamespace(MintMakerNamespaceName)
+			MaxSimultaneousPipelineRuns = 1
+
+			pplrName := "pplnr1"
+			labels := make(map[string]string)
+			labels[MintMakerAppstudioLabel] = "github"
+			pipelineRunBuilder := local_tekton.NewPipelineRunBuilder(pplrName, MintMakerNamespaceName)
+			pipelinerun, err := pipelineRunBuilder.WithLabels(labels).Build()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(k8sClient.Create(ctx, pipelinerun)).Should(Succeed())
+
+			pplrName = "pplnr2"
+			pipelineRunBuilder = local_tekton.NewPipelineRunBuilder(pplrName, MintMakerNamespaceName)
+			pipelinerun, err = pipelineRunBuilder.Build()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(k8sClient.Create(ctx, pipelinerun)).Should(Succeed())
+
+			Eventually(listPipelineRuns).WithArguments(MintMakerNamespaceName).Should(HaveLen(2))
+		})
+
+		_ = AfterEach(func() {
+			MaxSimultaneousPipelineRuns = originalMaxSimultaneousPipelineRuns
+
+			// Delete pipelineruns, so they don't leak to other tests
+			pipelineRuns := listPipelineRuns(MintMakerNamespaceName)
+			for _, pipelinerun := range pipelineRuns {
+				Expect(k8sClient.Delete(ctx, &pipelinerun)).Should(Succeed())
+			}
+		})
 
-		It("should successfully reconcile the resource", func() {
+		It("should launch 'github' pipelines first", func() {
 
-			// TODO(user): Add more specific assertions depending on your controller's reconciliation logic.
-			// Example: If you expect a certain status condition after reconciliation, verify it here.
+			existingPipelineRuns := tektonv1.PipelineRunList{
+				Items: listPipelineRuns(MintMakerNamespaceName),
+			}
+			var checkPipelineRun tektonv1.PipelineRun
+			for _, p := range existingPipelineRuns.Items {
+				if p.ObjectMeta.Name == "pplnr1" {
+					checkPipelineRun = p
+					break
+				}
+			}
+			// To count running pipelineruns, we have to check for empty 'status' fields;
+			// During the tests, these pipelineruns will not be updated to be actually 'running'
+			Expect(string(checkPipelineRun.Spec.Status)).To(Equal(""))
```

Comment:
This doesn't work for the purpose, you should check all github PipelineRuns should have start time later than non-github PipelineRuns. 

And actually if you check the PipelineRuns, you will see all have been started by the controller, for example with this test, it will fail:
```
pipelineRuns := listPipelineRuns(MintMakerNamespaceName)
for _, plr := range pipelineRuns {
        if plr.Labels[MintMakerAppstudioLabel] == "github" {
                Expect(plr.Spec.Status).To(BeEmpty())
        } else {
                Expect(plr.Spec.Status).To(Equal(tektonv1.PipelineRunSpecStatusPending))
        }
}
```

Reply:
Good catch! This testing logic is much more complete, thanks! I updated this test case with your suggestion.

When I ran it, it failed but it was due to a type mismatch. After I cast `plr.Spec.Status` to string, it worked.
I also included two boolean flags, to make sure both clauses were executed (and thus the non-github pipeline is indeed pending):
https://github.com/konflux-ci/mintmaker/blob/pipelinerun-controller-2/internal/controller/pipelinerun_controller_test.go#L115-L127

Do you think the new test is complete enough? And is is passing correctly for you as well?

Reply:
This still doesn't work for me because both the controller removed pending status from both github and non-github PipelineRuns. I believe the reason is in our testenv, there is actually no Tekton operator running for moving the PipelineRuns to "Running" status once the `PipelineRunSpecStatusPending` status is removed. So our controller always think there is no PipelineRUns are running there, thus it will "start" all PipelineRuns.

There are two possible solutions:
1. Implement a fake Tekton operator to start PipelineRuns (change the related fields `Status.Conditions.*` and etc)
2. Change the corresponding fields in PipelineRuns in our tests, for example:
```
BeforeEach(func() {
...
# max running num is 20
# create 20 PipelineRuns which are in "running" status.  Simulate the pipeline in a running state by setting the corresponding `Status.Conditions.*`  fields. 
# create 10 github PipelineRuns with "pending" status
# create 10 non-github PipelineRuns with "pending" status

It("should start github PipelineRuns first", func() {
...
# update 10 "running" PipelineRun to "completed" status.
# verify there are 20 PipelineRuns in "running" status again, and include 10 github PipelineRuns, all the 10 non-github PipelineRuns remain in "pending" status.
```
I believe the second solution is simpler and provides more flexibility for writing our tests.

Reply:
I'm fine with merging this PR now and address the test improvements later. If you proceed with merging, please add TODO/FIXME comments in the test file to indicate the needed improvements.

Reply:
+1 to merge this as it is. Also the other tests are all broken or incomplete. Let's just fix them all together and let's work on higher priority cards.

Reply:
Okay, I left a comment in that test case, and I'll proceed with merging this change as is. Thanks a lot for the input!
https://github.com/konflux-ci/mintmaker/pull/132/files#diff-450c910037cdd118f7ace46e459a73ccf20c3a163215ed441627c1cd5e51e2adR112-R113